### PR TITLE
Dp sf pap

### PR DESCRIPTION
**Pull request by daniellepace**
Improvements for 2D segmentation and pap project.

Includes:
- Conventional U-Net architecture
- Dice loss import and use
- Plotting Dice scores for comparisons
- Median computations for structures of interest, to .tsv and scatter plots
- 2D image data augmentation

Questions:
1. Includes changes to docker/vm_boot_images/config/tensorflow-requirements.txt. Shall we push these everywhere?
2. Remaining issues with explorations/infer_medians, around line 740, as I refactored it to remove hard-coded items. Previously, I ran recipes with no --output-tensors if I wanted to do inference on all subjects. However, I now need the tensor_maps_out for its channel_map, output_name, etc. I tried removing the output tensors when I call test_train_valid_tensor_generators so that I don't filter on images with segmentations, and instead run inference on all subjects, but the code still gives me a .tsv with only ~500 subjects and not ~60K.

### CODE REVIEW

File: ml4h/arguments.py [hunk 1]
```diff
@@ -269,6 +272,11 @@ def parse_args():
         help='If true saves the model weights from the last training epoch, otherwise the model with best validation loss is saved.',
     )
 
+    # 2D image data augmentation parameters
+    parser.add_argument('--rotation_factor', default=0., type=float, help='a float represented as fraction of 2 Pi, e.g., rotation_factor = 0.014 results in an output rotated by a random amount in the range [-5 degrees, 5 degrees]')
```
Comment (lucidtronix) [inline ml4h/arguments.py:276]: Add that these are data augmentation specific to the help strings, eg "For data augmentation, a float....'

File: ml4h/arguments.py [hunk 2]
```diff
@@ -376,6 +384,16 @@ def parse_args():
         default='3M',
     )
 
+    # Arguments for explorations/infer_medians
+    parser.add_argument('--analyze_ground_truth', default=True, help='Whether or not to filter by images with ground truth segmentations, for comparison')
+    parser.add_argument('--dates_file', help='File containing dates for each sample_id')
```
Comment (lucidtronix) [inline ml4h/arguments.py:389]: Can we re-use the more general existing arg app_csv for this? Just trying to mitigate argument explosion.

File: ml4h/arguments.py [hunk 3]
```diff
@@ -376,6 +384,16 @@ def parse_args():
...
+    parser.add_argument('--structures_to_analyze', nargs='*', default=[], help='Structure names to include in the .tsv outputs')
+    parser.add_argument('--erosion_radius', default=1, type=int, help='Radius of the unit disk structuring element for erosion preprocessing')
+    parser.add_argument('--intensity_thresh', type=float, help='Threshold value for preprocessing')
+    parser.add_argument('--intensity_thresh_in_structures', nargs='*', default=[], help='Structure names whose pixels should be replaced if the images has intensity above the threshold')
+    parser.add_argument('--intensity_thresh_out_structure', help='Replacement structure name')
+    parser.add_argument('--results_to_plot', nargs='*', default=[], help='Structure names to make scatter plots')
```
Comment (lucidtronix) [inline ml4h/arguments.py:395]: Can we re-use structures_to_analyze for this?

File: ml4h/explorations.py [hunk 4]
```diff
@@ -691,6 +699,191 @@ def infer_with_pixels(args):
...
+
+def _thresh_labels_above(y, img, intensity_thresh, in_labels, out_label, nb_orig_channels):
+    y = np.argmax(y, axis=-1)[..., np.newaxis]
+    y[np.logical_and(img >= intensity_thresh, np.isin(y, in_labels))] = out_label
+    y = y[..., 0]
+    y = _to_categorical(y, nb_orig_channels)
+    return y
+
+def infer_medians(args):
+    assert(args.batch_size == 1) # no support here for iterating over larger batches
```
Comment (lucidtronix) [inline ml4h/explorations.py:730]: as this is "public" add a Docstring

Comment (lucidtronix) [reply ml4h/explorations.py:730]: Also maybe rename to make it a little more specific maybe infer_medians_from_segmented_regions or something...

File: ml4h/explorations.py [hunk 5]
```diff
@@ -691,6 +699,191 @@ def infer_with_pixels(args):
...
+    assert(len(args.tensor_maps_in) == 1) # no support here for multiple input maps
```
Comment (lucidtronix) [inline ml4h/explorations.py:731]: You can bake the comment in to the assert with `assert(len(args.tensor_maps_in) == 1, 'no support here for multiple input maps')` and then it is a bit easier to debug.

File: ml4h/explorations.py [hunk 6]
```diff
@@ -691,6 +699,191 @@ def infer_with_pixels(args):
...
+    assert(len(args.tensor_maps_out) == 1) # no support here for multiple output channels
+
+    tm_in = args.tensor_maps_in[0]
+    tm_out = args.tensor_maps_out[0]
+    assert(tm_in.shape[-1] == 1)  # no support here for stats on multiple input channels
+
+    # don't filter datasets for ground truth segmentations if we want to run inference on everything
+    # TODO HELP - this isn't giving me all 56K anymore
```
Comment (lucidtronix) [inline ml4h/explorations.py:739]: Does this need to be addressed or can wait till re-ingest?

File: ml4h/explorations.py [hunk 7]
```diff
@@ -691,6 +699,191 @@ def infer_with_pixels(args):
...
+            inference_writer_pred.writerow(csv_row_pred)
+
+            tensor_paths_inferred.add(tensor_paths[0])
+            stats_counter['count'] += 1
+            if stats_counter['count'] % 250 == 0:
+                logging.info(f"Wrote:{stats_counter['count']} rows of inference.  Last tensor:{tensor_paths[0]}")
+
+    # Scatter plots
+    if args.analyze_ground_truth:
+        df_true = pd.read_csv(inference_tsv_true, delimiter='\t', quotechar='"', quoting=csv.QUOTE_MINIMAL)
```
Comment (lucidtronix) [inline ml4h/explorations.py:835]: refactor this into a function just to keep things modular, arguably it belongs in `plots.py` but only if you think we might use it in another context.

File: ml4h/metrics.py [hunk 8]
```diff
@@ -260,6 +262,24 @@ def loss(y_true, y_pred):
...
+    dice_fxns = []
+    for label_key in labels:
+        label_idx = labels[label_key]
+        fxn_name = label_key.replace('-', '_').replace(' ', '_')
+        string_fxn = 'def ' + fxn_name + '_dice(y_true, y_pred):\n'
+        string_fxn += '\tdice = Dice(laplace_smoothing=1e-05).dice(y_true, y_pred)\n'
+        string_fxn += '\tdice = K.mean(dice, axis=0)['+str(label_idx)+']\n'
+        string_fxn += '\treturn dice'
+
+        exec(string_fxn)
```
Comment (lucidtronix) [inline ml4h/metrics.py:278]: <3 great job figuring out this tricky code pattern :)

File: ml4h/plots.py [hunk 9]
```diff
@@ -2733,6 +2737,93 @@ def plot_precision_recalls(predictions, truth, labels, title, prefix="./figures/
     plt.savefig(figure_path)
     logging.info("Saved Precision Recall curve at: {}".format(figure_path))
 
+def plot_dice(predictions, truth, labels, paths, title, prefix="./figures/", dpi=300, width=3, height=3):
+    label_names = labels.keys()
```
Comment (lucidtronix) [inline ml4h/plots.py:2741]: Add Docstring and ideally type hints, though we have not done a great job of that in general.

File: ml4h/tensor_generators.py [hunk 10]
```diff
@@ -666,13 +666,6 @@ def get_train_valid_test_paths(
 
     logging.info(f'Found {len(train_paths)} train, {len(valid_paths)} validation, and {len(test_paths)} testing tensors at: {tensors}')
     logging.debug(f'Discarded {len(discard_paths)} tensors due to given ratios')
-    if len(train_paths) == 0 or len(valid_paths) == 0 or len(test_paths) == 0:
```
Comment (lucidtronix) [inline ml4h/tensor_generators.py:669]: can we just replace the `or` with `and` if there is no data we still want fail fast right?

File: ml4h/tensor_generators.py [hunk 11]
```diff
@@ -732,6 +725,51 @@ def get_train_valid_test_paths_split_by_csvs(
             logging.info(f"CSV:{balance_csvs[i-1]}\nhas: {len(train_paths[i])} train, {len(valid_paths[i])} valid, {len(test_paths[i])} test tensors.")
     return train_paths, valid_paths, test_paths
 
+# https://stackoverflow.com/questions/65475057/keras-data-augmentation-pipeline-for-image-segmentation-dataset-image-and-mask
+def augment_using_layers(images, mask, in_shapes, out_shapes, rotation_factor, zoom_factor, translation_factor):
+
```
Comment (lucidtronix) [inline ml4h/tensor_generators.py:730]: Add Docstring and typehints

File: ml4h/tensor_generators.py [hunk 12]
```diff
@@ -828,14 +875,31 @@ def test_train_valid_tensor_generators(
         paths=test_paths, num_workers=num_train_workers, cache_size=0, weights=weights,
         keep_paths=keep_paths or keep_paths_test, mixup_alpha=0, name='test_worker', siamese=siamese, augment=False,
     )
-    if wrap_with_tf_dataset:
+
+    do_augmentation = bool(rotation_factor or zoom_factor or translation_factor)
+    logging.info(f'doing_augmentation {do_augmentation}')
```
Comment (lucidtronix) [inline ml4h/tensor_generators.py:880]: It's already in the log at the top with all of the arguments so this not needed, but if you want to keep the info statement here I would add the actual values of the augmentation arguments.

File: ml4h/tensor_generators.py [hunk 13]
```diff
@@ -828,14 +875,31 @@ def test_train_valid_tensor_generators(
...
+
+    if do_augmentation:
+        assert(len(tensor_maps_in) == 1) # no support for multiple input tensors
```
Comment (lucidtronix) [inline ml4h/tensor_generators.py:883]: Bake comments into asserts

File: scripts/jupyter.sh [hunk 14]
```diff
@@ -15,6 +15,7 @@ DOCKER_COMMAND="docker"
 PORT="8888"
 SCRIPT_NAME=$( echo $0 | sed 's#.*/##g' )
 GPU_DEVICE="--gpus all"
+ENV=""
```
Comment (lucidtronix) [inline scripts/jupyter.sh:18]: I'm curious what you use this for?

**Review by lucidtronix — CHANGES_REQUESTED** [review]
@daniellepace This is tremendous! thank you for the excellent additions! All my comments are small aesthetic things mostly around documentation.

File: scripts/jupyter.sh (hunk 14 above)
Comment (daniellepace) [reply scripts/jupyter.sh:18]: I use this if I want to use a notebook that uses ml4h code that I am actively developing. I can set my PYTHONPATH to my ml4h directory so that I can edit the ml4h code and have it accessible to the notebook. I use something similar when I run docker so that I can edit the ml4h code and have it show up in the docker without having to restart it.

File: ml4h/explorations.py (hunk 7 above)
Comment (daniellepace) [reply ml4h/explorations.py:835]: I refactored this into its own function, and agree it looks much cleaner now. But, it's very specific to this context so for now I think we can keep it in explorations.py

File: ml4h/explorations.py (hunk 6 above)
Comment (daniellepace) [reply ml4h/explorations.py:801]: It's up to us - the code doesn't currently work as it should right now. If args.analyze_ground_truth is True, then it should work on 56K tensors, but right now it still filters for only tensors that have ground truth. I had tried setting args.output_tensors and args.tensor_maps_out to empty lists in an attempt to override the code that filters for output tensors, but it's not currently working. That said, the project is in a waiting phase anyways now.

**Review by lucidtronix — APPROVED** [review]
Awesome work!
